stdlib/list.ml

```diff
@@ -125,6 +119,23 @@ let rec fold_right f l accu =
     [] -> accu
   | a::l -> f a (fold_right f l accu)
 
+let rev_flatten l =
+  fold_left (fun accu l -> rev_append l accu) [] l
+
+let rec flatten_aux len l =
+  match l with
+    [] -> []
+  | l1 :: r ->
+    let new_len = len + length l1 in
```

Comment:
Computing the length means that all lists will be traversed twice.  You should keep track of the number of elements while traversing `l1` during concatenation.